### ISSUE TITLE
Support explicit braces 

### ISSUE THREAD

**Issue by paf31** [purescript/purescript#774]
In Haskell, braces can be used instead of indentation. This is useful in some circumstances, and I think it would be nice to support it.

**Comment by garyb** [comment]
Continues from the other issue: apart from `psci` cases, what other sort of situations do you think this is useful?

I think the fact we have one consistent kind of delimiter is far better than adding an alternate (and less elegant) way of doing the same thing. I consider us _not_ having braces and semicolons a definite improvement over Haskell.

**Comment by paf31** [comment]
I just find them useful sometimes in cases where the indentation gets hard to parse. In that case, `{` and `;` are just a little more explicit. I don't want to go "full GHC" and use the feature in large blocks everywhere, but I do like them occasionally with `do` notation, especially when writing parsers.

There is an issue of ambiguity, due to `{` also being used in record binders.

I'm not 100% sure I want this yet. Just wanted to add my 2¢.

**Comment by user471** [comment]
>what other sort of situations 

What about using do notation as list comprehension replacement?
```
do x <- 0..2; y <- 0..2; pure $ x + y
```

**Comment by jdegoes** [comment]
It'd be nice to get *just* the semi-colon for one-line `do` / `let` but without the hideous curly braces. Nonetheless, nothing wrong with just having one way to do it, makes code syntax more uniform, if slightly more verbose in some cases.

**Comment by paf31** [comment]
@jdegoes Sounds interesting, if I can figure out how to make indentation work between semicolons.

**Comment by user471** [comment]
Could be useful with Prisms
```
_Foo = prism Foo (case x of (Foo x) = Right x; y = Left y)
```
similar to lens
```
foo = lens _.foo (_ { foo = _ })
```

**Comment by AppShipIt** [comment]
I've been noodling this because it's relavamt to my comma proposal. @paf31 Do you have an example of <q>hard to parse </q> indentation? Although I principally agree with @jdegoes about <q>one way to do it</q> my strong opposition is providing syntax to write less readable code. @user471's comprehension is exemplar. It's difficult to glance over relative to the multiline alternative (and it's simpler than what I really write). I suspect when indentation gets asty there needs no be more granularity in a `where` or `let` clause.

**Comment by AppShipIt** [comment]
For context Elm endorses style conventions <q>for humans</q> I found annoying at first but really do ensure write-to-be-read code. Multiple lines in one is pretty much the opposite or even braces enabling people to individual indentation schemes.

**Comment by garyb** [comment]
@AppShipIt just to clarify, you're saying that this is a bad idea? I read it the other way around at first, I thought you were approving of it :wink:

**Comment by AppShipIt** [comment]
Lol sorry! Bad idea from me because I'd prefer to read properly formatted code.

**Comment by user471** [comment]
@AppShipIt @garyb 
So you think that
```
_Foo = prism Foo (case x of 
                   (Foo x) = Right x; 
                   y = Left y)
```
better than
```
_Foo = prism Foo (case x of (Foo x) = Right x; y = Left y)
```
?

**Comment by garyb** [comment]
No, I'd use a `where`. But also I don't think semicolons wouldn't work in the case example, they're for use in `do` blocks - using them to terminate cases too would be ambiguous.

If I didn't want to use a `where` even this is better:

``` purescript
_Foo = prism Foo case x of 
  Foo x -> Right x
  y -> Left y
```

(There's no need for parens in that example).

**Comment by zudov** [comment]
I am not sure whether I like the semicolon separated `do`-notation. If one has a very tiny `do` block which could be squeezed in one liner, why not to just use `bind` and `apply`? If literal `bind` and `apply` are too noisy in some particular case I personally find newline separated `do`-notation much easier to read.

Pretty much the same applies to `case`, squashing lines with semicolons would save you a few lines at most, but I hardly see how that is a benefit.

Having alternative non-indentation-sensitive syntax with braces and semicolons is indeed useful, but Haskell shows that nobody really uses it, and when someone does it [spawns](https://www.reddit.com/r/haskell/comments/1fqyrk/pixel_art_editor_in_600_lines_of_haskell/cad7ya8) [questions](https://www.reddit.com/r/programming/comments/1fo3c0/pixel_art_editor_in_600_lines_of_haskell/cacex28).

**Comment by phadej** [comment]
Haskell has `do { x <- action1; action2 x }` and `case exp of { A a -> withA a; B b -> withB b }` syntax.

It's handy for machine-generating stuff.

Syntactically important thing is that, the blocks are started `{`, and ended `}`, so you can nest `do` and `case`. Also parser definition is quite simple.

```hs
do { x <- action1; case x of { [] -> action2; (x:xs) -> action3 x xs } }
```

But in PureScript curly-braces have found a better use already (so using them will introduce another ambiguity). Yet without "start a context for semicolon separated expressions" -token, the grammar becomes tricky to formulate.

**Comment by AppShipIt** [comment]
> Said that, it would be very nice to have semicolons and braces in psci, but probably it would be super awkward if some syntax is supported only by psci. 

@zudov Sorry I haven't played with interactive mode yet. Wouldn't it be better to handle multiple lines?

**Comment by zudov** [comment]
@AppShipIt Multiline input would be nice, but aligning indentation is so not handy when in REPL. Or probably it's just that my readline/haskelline skills are poor and I am not aware a bunch of keybindings that make it handy.

**Comment by hdgarrood** [comment]
Yes, I'd be tempted to close this issue at this point.

**Comment by natefaubion** [comment]
I think we should close it. We have case sections and `:paste` mode, and this ticket hasn't seen any progress in nearly 3 years.

**Comment by hdgarrood** [comment]
Sorry, what are case sections?

**Comment by natefaubion** [comment]
`case _ of`, in reference to the `prism` example.
